### DISCUSSION TITLE
Allow mixing "init;" and "private set;"

### DISCUSSION THREAD

**Discussion by lucasteles** [dotnet/csharplang#7275] [Language Ideas]
Sometimes we want our class public API to be immutable as possible and delegate all state management to the class itself.
The combination of `required` and `init` is very handy on this, but today it is not easy to have an `init` property with `private set;`/`protected set';`

Would be nice to have a simple way to do this.

Some ideas:
```csharp
public class Foo1
{
    public required string Value { get; init and private set; }
    public void Update(string value) => Value = value;
}
public class Foo2
{
    public required string Value { get; init with private set; }
    public void Update(string value) => Value = value;
}
public class Foo3
{
   // just allow both with some precedence
    public required string Value { get; init private set; }
    public void Update(string value) => Value = value;
}
public class Foo4
{
    public required string Value { get; init, private set; }
    public void Update(string value) => Value = value;
}
public class Foo5
{
    [AllowSet(Visibility.Private)]
    public required string Value { get; init; }
    public void Update(string value) => Value = value;
}
```

When mixing `init` and `set`  should have a constraint that the `set`  must be less accessible than `init`

**Comment by HaloFour** [comment]
The `init` accessor method is really the `set` accessor method with additional metadata applied to it via `modreq` to indicate when/where it could be called.  I don't think you could have a separate accessor method here, but you could potentially have the metadata inform the compiler that the method could be called in other circumstances.  For a `private set` scenario that should be easier since it's in the same compilation unit and does not need to impact the external metadata of the accessor.

**Reply by lucasteles** [reply]
Maybe an attribute?

**Comment by TahirAhmadov** [comment]
Can you cite an example of how these classes would be used?

**Reply by brendonparker** [reply]
My use case would be for EF where I want to insulate the entity such that state can only be changed from within (i.e. `private set;`), but I also want to be able to initialize the properties. So, the ability to initialize state to whatever (for purpose of saving to database), but then allowing the entity to restrict what changes are made after the fact (for writes going forward).

Perhaps combined with `record` types, this gives some flexiblity, especially around testing. For an immutable record type, can't modify the state of the object, but you _can_ create a copy (using the `with`) syntax. This makes setting tests up easier. Its a bit of a lazy approach, I suppose, as you're working around needing to fully setup the object and go through business logic. 

Concrete (yet contrived) example:

```
public record Order
{
    public required long Id { get; init; }
    public required string Status { get; init private set; }
    // Other fields here that may be related to Status ...
    
    public void StartProcessing()
    {
        Status = "Processing";
        // More logic here touching other properties,
        // Requiring certain state to be setup
    }
}

Order order = new() { Id = 0, Status = "Queued" };
order.StartProcessing();

// Ideally, could also do this, to get the same order, but with the Status set to Processed:
Order order2 = order with { Status = "Processed" };

order2.Status.Should().Be("Processed");
```

**Comment by DavidArno** [comment]
Since the following is already possible:
```cs
public class Foo1
{
    private string _value;
    
    public required string Value 
    { 
        get => _value; 
        init => _value = value; 
    }

    public void Update(string value) => _value = value;
}
```
I'd question the value of this proposal. The occasions where you want public init, private set and provide a public method to work around that public init seem likely to be vanishingly rare. As @TahirAhmadov asks, do you have a real-world use case for where having a public `set` causes a problem which this proposal then solves?

**Reply by FaustVX** [reply]
Going from auto-property to full property just to have a private set is anoying (it's one of the reason the `field` keyword will be introduced).
Also, if you want to have some validation logic you have to also write a `SetValue()` method.

**Reply by DavidArno** [reply]
@FaustVX, the question still stands: what is the problem that this is trying to solve? What's wrong with just having a public `set`, especially when we get the `field` keyword?

**Reply by FaustVX** [reply]
> What's wrong with just having a public `set`,

Because we don't want the outside to modify the value after it's creation, but still want the possibility to modify it inside the class.
I have already wanted this feature some time ago, but I can't remember what it was.

**Reply by lucasteles** [reply]
@FaustVX exactly

**Reply by lucasteles** [reply]
@DavidArno The compiler will emit a warning on your code snippet also: `Non-nullable field '_value' is uninitialized. Consider declaring the field as nullable.`

**Reply by 333fred** [reply]
Only a minor modification needed:

```cs
public class Foo1
{
    private string _value;
    
    public required string Value 
    { 
        get => _value; 
        [MemberNotNull(nameof(value))]
        init => _value = value; 
    }

    public void Update(string value) => _value = value;
}

**Reply by Bretttt** [reply]
@DavidArno In my opinion, the introduction of the `field` keyword doesn't weaken the case for the proposal, as your example becomes something like

```
public class Foo1
{    
    public string Value 
    { 
        get => field; 
        init, internal set
        {
            if (value != field)
            {
                // validation, etc...
                field = value;
            }
        }
    }
}
```

**Reply by jnm2** [reply]
IIRC we even discussed having both `init` and `set` at some point when brainstorming `field` and how to get nullability right with constructor initialization without too much analysis magic.
Just a note, I'd expect the sample to be formatted this way:
```cs
public string Value 
{ 
    get; 
    init;
    internal set
    {
        if (value != field)
        {
            // validation, etc...
            field = value;
        }
    }
}
```

**Comment by TheBrambleShark** [comment]
Question on the IL side - I see three possible implementation paths here. This assumes we are using the following syntax:
```cs
public int Foo { get; init; set; }
```

1. Introduce `.init`. Breaking change, but it would mean that init properties are now truly readonly rather than conditionally settable.
```cs
// Properties
.property instance int32 Foo()
{
    .get instance int32 C::get_Foo()
    // .set instance void modreq([System.Runtime]System.Runtime.CompilerServices.IsExternalInit) C::init_Foo(int32)
    .init instance void C::init_Foo(int32)
    .set instance void C::set_Foo(int32)
}
```

2. Use `.set` as we are now, but multiple entries.
```cs
// Properties
.property instance int32 Foo()
{
    .get instance int32 C::get_Foo()
    .set instance void modreq([System.Runtime]System.Runtime.CompilerServices.IsExternalInit) C::init_Foo(int32)
    .set instance void C::set_Foo(int32)
}
```

3. Use `.other`, which is already reserved for alternative uses.
```
// Properties
.property instance int32 Foo()
{
    .get instance int32 C::get_Foo()
    .other instance void modreq([System.Runtime]System.Runtime.CompilerServices.IsExternalInit) C::init_Foo(int32)
    .set instance void C::set_Foo(int32)
}
```

I think allowing us to emit multiple `.set` instructions (not sure what the name of the keyword here is) would be the best option, since it's backwards compatible, but I'm not entirely sure how the system would handle navigating which one to choose. We would need almost some sort of overload resolution on property setters (perhaps this already exists).

On the method side, we'd get all of the standard members exactly how we'd expect (bodies omitted).
```cs
.method public hidebysig specialname instance int32 get_Foo () cil managed
.method public hidebysig specialname instance void modreq([System.Runtime]System.Runtime.CompilerServices.IsExternalInit) init_Foo (int32 'value') cil managed 
.method public hidebysig specialname instance void set_Foo (int32 'value') cil managed
```

I suspect the defaults would be the IL equivalent of this:
```cs
public int Foo
{
    get => field;
    init => field = value;
    set => field = value;
}
```

Other open questions:
1. Should `init` forward to `set` when the body of `set` is not defined (i.e. auto property)
2. If a body for `set` is defined, should `init;` be permitted (defaulting to `init => field = value;` or should it be required that `init` is also explicitly defined in this case?
3. What visibility modifiers should we allow on `set` - should it be restricted to non-public or are there use cases for public setters (e.g. init bypasses validation imposed on set)?
4. If we choose `.init`, how is `Foo { get; init; }` handled - same as today (`.set` with `modreq`) or should we always emit `.init`?
5. If we choose `.init`, would we need the modreq for `init_Foo` or can that be implied by the `.init` keyword?
6. If we choose `.init`, should we produce both `.init` and `.set` with `modreq` for backwards compatibility when multi-targeting previous versions?

**Reply by colejohnson66** [reply]
It was very much a design discussion when initializers were added about whether to have a `.init` accessor. It was decided that `modreq` on `.set` was the better option.

**Comment by CaiB** [comment]
I support this request.

My usecase is the following:
I have a class that represents an item in a database. I want to be able to instantiate the class as part of creating a new entry in the database, and ensure that the class is completely filled with valid values. I accomplish this with use of the `required` keyword on most `public ... { get; init; }` properties, allowing compile-time checking to ensure all necessary data is set, plus additional verification before submission to the database driver for validity testing of the set values. I specifically only expose a `public init`, as setting the property from outside the class anytime other than init would create a disconnect between the class and database state.
Database updates are implemented as async instance methods on the class. AFAIK async is not possible with property setters, and I strongly believe in not doing time-intensive work inside of property getters/setters, and a network round-trip to a database certainly falls under that. It also does not allow me to return whether the operation succeeded, or an error message if not, without having the property setter throw an exception (which also goes counter to how one would expect properties to function).
Hence, the most logical thing would be to have a convenient way to add `private set` semantics to the properties (i.e. `{ get; init; private set; }`), so that once a database update is successfully done, the class can internally mutate its state to once again match the database state, without losing external direct immutability.
